torch/distributed/_checkpoint/metadata.py

```diff
@@ -0,0 +1,47 @@
+from typing import Any, Dict, List, Union, Optional
+import torch
+from dataclasses import dataclass
+import pickle
+from torch.distributed.shard.sharded_tensor import ShardedTensor, ShardedTensorMetadata, ShardMetadata
+
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class StorageMetadata:
+    shard_metadata: Optional[ShardMetadata]
+    # Unique identifier for this particular entity (Tensor or Shard of ShardedTensor)
+    storage_key: str
+    length: int
+    offset: int
+
+# Metadata for each param.
+@dataclass
+class ExtendedTensorMetadata:
+    # Details of Tensor/ShardedTensor (dtype, shape, sharding config etc.)
+    tensor_metadata: ShardedTensorMetadata
```

Comment:
If we have a regular `torch.Tensor`, how do we represent its metadata here?

Reply:
I would think there need be an equivalent to ShardedTensorMetadata, let me add some inline comment